Add function to clean up column names with special characters

hwalinga opened the pull request:
Changed the backtick quoting functions so that you can also use backtick quoting to use invalid Python identifiers like ones that start with a digit, start with a number, or are separated by operators instead of spaces. 

- [x] closes #27017
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

As this builds upon #24955 I think @jreback would be again the right person for the code review.

hwalinga commented:
I don't know how I can solve this issue `pandas/core/computation/common.py:4: error: Module 'tokenize' has no attribute 'EXACT_TOKEN_TYPES'` as the tokenize module has this attribute. (In the "Type validation" check.)

simonjayhawkins commented:
https://docs.python.org/3/library/tokenize.html#module-tokenize does not refer to EXACT_TOKEN_TYPES and typeshed does not have a definition for EXACT_TOKEN_TYPES https://github.com/python/typeshed/blob/master/stdlib/3/tokenize.pyi

my guess is it's not part of the public API for tokenize.

hwalinga commented:
@simonjayhawkins I see. Shall I write out this dictionary instead of importing it?

simonjayhawkins commented:
just reading the docs.. "All constants from the token module are also exported from tokenize." https://docs.python.org/3/library/tokenize.html#tokenizing-input

so it would perhaps appear that the omission of EXACT_TOKEN_TYPES is an oversight in typeshed.

so probably best to add `# type: ignore` and a comment.

simonjayhawkins commented:
> just reading the docs.. "All constants from the token module are also exported from tokenize." https://docs.python.org/3/library/tokenize.html#tokenizing-input
> 
> so it would perhaps appear that the omission of EXACT_TOKEN_TYPES is an oversight in typeshed.
> 
> so probably best to add `# type: ignore` and a comment.

forget that. I misread the docs, 

from the source..

```
import token
__all__ = token.__all__ + ["tokenize", "detect_encoding",
                           "untokenize", "TokenInfo"]
```

simonjayhawkins commented:
can you use tokenize.tokenize instead? The returned named tuple has an additional property named exact_type that contains the exact operator type for OP tokens. For all other token types exact_type equals the named tuple type field.

hwalinga commented:
@simonjayhawkins 
Well, I basically need a list of the operators first, and than I need the number associated with that operator. I actually don't know how exactly I would do that like that. 

So, I guess, I either ignore, or write them out manually.

WillAyd commented on pandas/core/frame.py:
```diff
@@ -3137,8 +3137,13 @@ def query(self, expr, inplace=False, **kwargs):
+            names that start with a digit, or are a Python keyword. Basically
+            when it is not valid Python identifier.
+
+            NB. You cannot use this if there are multiple invalid characters
```
Is there a practical reason for this limitation? Seems a little arbitrary

hwalinga replied:
Good catch, this was sloppy. This is not true.

However, there is an edge case here, it is a bit more complicated. I could replace it too: 

            NB. A combination of an invalid character with a space, or
            multiple spaces will collapse the space. The reason is that the
            string will be parsed as Python code
            and in Python ``a*b`` and ``a* b`` are equal.
            So, if you want to refer to a column called ``"a* b"`` you
            can use ```a*b```, but it is recommend in this case to change the
            column name or use bracket notation, to prevent very opaque code.

But, since this is such an edge case, it would only make the documentation confusing. Just leave it out?

WillAyd replied:
So IIUC here the danger is that valid expressions will be evaluated? Or is this a matter of how we are handling whitespace in this PR?

hwalinga replied:
No, it isn't a matter of expressions being evaluated, it is a matter of expressions being parsed, and they are parsed as Python code, and therefore whitespace sometimes disappears.

Let's just say, there are some edge cases (which cannot be solved easily), but they are rare. Should we still mention them somewhere?

hwalinga replied:
Can you see if it can be merged now?

hwalinga replied:
This edge case is now largely resolved. (Although, one can still construct certain query + column names combination to confuse the parser, but such cases are much rarer now.)

hwalinga replied:
I could maybe see if I can make code that detects such edge cases and raises an appropriate exception. (Now they will go by silent, and the result may be false or correct.)

hwalinga replied:
Done.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -488,16 +488,16 @@ def _get_index_resolvers(self):
 
-    def _get_space_character_free_column_resolvers(self):
-        """Return the space character free column resolvers of a dataframe.
+    def _get_special_character_free_column_resolvers(self):
```
Slightly orthogonal to this PR but this function only works for DataFrames right? If so might make sense to move from generic to DataFrame (unless it can actually operate on a Series)

hwalinga replied:
Technically yes, it can operate on Series. However, since eval/query are only methods available for DataFrames, it will never be used on a Series by a library function unless someones explicitly calls it. 

So, shall I move it?

WillAyd replied:
Keep here for now can move in a follow up

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -40,6 +40,24 @@
 import pandas.io.formats.printing as printing
 
 
+def tokenize_backtick_quoted_string(token_generator):
```
Can you add annotations for new functions?

hwalinga replied:
I did for the simple types, but a `token_generator` is a `Generator[Tuple[int, str, Tuple[int, int], Tuple[int, int], str], None, None]`, and I wonder how helpful this is, or that this is just a lot of noise.

WillAyd replied:
I think you should be able to match what's in type shed for the function, so `Generator[TokenInfo, None, None]`

hwalinga replied:
I think I understand.

hwalinga commented:
Changed code so that the PR now also closes #28576.

hwalinga commented:
Also, I moved some more functions from pandas/core/computation/expr.py to pandas/core/computation/common.py. But the latter isn't anymore a good place for that. It is probably better to have them moved all to expr.py again, but maybe it is also a good idea to put them in a new file backtickquoting.py? (About 100 lines)

hwalinga commented:
Implemented code that raises an exception if there is ambiguity. (Ambiguity should be very rare.)

Also added an exception if there somehow exists a way the conversion fails to create a valid Python identifier. (Currently cannot think of one.)

Also moved some functions around.

Probably need a new code review. @WillAyd you want to take a look again?

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -39,8 +36,69 @@
+    # By my knowlegde the current conversion cannot fail, but who knows.
+    if not name.isidentifier():
+        raise ValueError(
+            "Could not convert %s to a valid Python identifier" % name +
```
Can you use `.format` instead of Py2 syntax for string replacement?

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -39,8 +36,69 @@
+    name = "".join(special_characters_replacements.get(char, char) for char in name)
+    name = "BACKTICK_QUOTED_STRING_" + name
+
+    # By my knowlegde the current conversion cannot fail, but who knows.
```
Hmm do we have a test case that hits this?

hwalinga replied:
Well, I just cannot imagine something that could go wrong. We have unit tests that test things what came to mind, but they work. I cannot come up with a test case that tests if it would fail, because than I probably made a workaround for this. This exception is more there for the case that someone in the future by accident finds such a case, and then he/she gets a readable exception telling that raising an issue can help us improve this feature.

WillAyd replied:
I guess if we can't think of a way to actually hit the code branch we should just remove; speculating that this could get hit doesn't help too much

hwalinga replied:
I see. 

However, I found some way this can be hit. 

You can for example use emoticons as column names. There is now an test case that tests this to raise an exception. 

I don't think I will come up with something that can actually work with emoticons as column names. So, let's rest that for now. 

(I might be able come up with something, but let's just say that we just shouldn't go into such craziness.)

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -39,8 +36,69 @@
+
+
+def _tokenize_backtick_quoted_string(
+    token_generator: Generator[tokenize.TokenInfo, None, None]
```
```suggestion
    token_generator: Iterable[tokenize.TokenInfo]
```

I think the Python docs suggest doing this when there is no send or return type for the `Generator` and would be more readable

WillAyd commented on pandas/core/generic.py:
```diff
@@ -489,16 +489,43 @@ def _get_index_resolvers(self):
+            for k, v in self.items()
+        ]
+
+        cleaned_name_count = collections.Counter(map(operator.itemgetter(0), resolvers))
```
Can you add a comment as to the purpose of this? Would help future readers

WillAyd commented on pandas/core/generic.py:
```diff
@@ -489,16 +489,43 @@ def _get_index_resolvers(self):
+        for name in backtick_quoted_variables:
+            if cleaned_name_count[name] > 1:
+                raise ValueError(
+                    "Description %s is ambiguous for the column names." % name +
```
Same comment on using `.format` instead of `%s`

WillAyd reviewed (changes requested):
OK thanks. I think moving in the right direction just hoping we can improve readability

hwalinga commented:
@WillAyd What do you think of it now. (I am not entirely sure why travis fails, it says something about TestS3, so maybe a (temporal) problem with AWS?)

WillAyd commented on pandas/tests/groupby/test_categorical.py:
```diff
@@ -782,7 +782,7 @@ def test_categorical_no_compress():
 def test_sort():
 
-    # http://stackoverflow.com/questions/23814368/sorting-pandas-categorical-labels-after-groupby  # noqa: flake8
+    # http://stackoverflow.com/questions/23814368/sorting-pandas-categorical-labels-after-groupby  # noqa
```
What was this change for?

hwalinga replied:
Somehow flake8 started whining about this line. It is an incorrect way of ignoring a line. I guess a flake8 update made this kind of indicating an ignoring illegal. So I fixed it to make flake8 pass. 

http://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -39,8 +36,74 @@
+        return name
+
+    # Create a list with the special characters and their replacement.
+    # So far we only replace single character operators.
```
Why is this restriction in place? Not trying to pester just want to be as clear as possible on where to draw the line if making changes

hwalinga replied:
That's a wrong comment indeed. Multiple character operators are composed of single operators, and therefore automatically replaced. I revisit these comments and made them clearer.

hwalinga commented:
@WillAyd What do you think. Is it ready already, and should I maybe to a squash or something?

WillAyd commented on pandas/core/generic.py:
```diff
@@ -489,16 +489,51 @@ def _get_index_resolvers(self):
+            _tokenize_string,
+        )
+
+        resolvers = [
```
Hmm so what is the point of converting `self.items()` to a list of tuples to ultimately convert back to a dict? IIUC you are just trying to detect if the cleansed names clash and get duplicated right? If so why not keep this a dict and just compare the length to `self.items()` after the fact?

hwalinga replied:
I need to have `resolvers` as a list of tuples as I am actually wanting to detect the exact name that causes such a clash. 

I do this in the next code line by creating `cleaned_name_count`, a Counter object containing counts for the cleansed names. If there is a cleansed name with count > 1, we know it clashed an I throw an error. 

Is this a problem with performance? If so, I can do as you recommend and only start the procedure to find the clash when the dicts have different length.

WillAyd replied:
Not a performance issue just readability. Wouldn't taking the set difference of the dict keys before and after give you the same information?

hwalinga replied:
No, the cleansing changes the keys of the dictionary. The values stay the same though, but hashing the values would really be a performance issue. 

After thinking about it, I do think your proposed method is a more readable approach. It really looks a lot more like a guard clause that way. 

But I don't think there is nice way to do the check elegantly, and get the clashed names. You just have to redo the conversion.

Also, another complicated factor. I only throw the error if the expr uses a name from one of the columns that clashes. So, if the dataframe has clashing column names, but the expr does not uses one of them. It is not a problem and the error is not raised. 

Okay, I am gonna try something, and you can say if this is an improvement or not.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -489,16 +489,51 @@ def _get_index_resolvers(self):
+        resolvers = [
+            (
+                _clean_special_characters_column_name(
+                    next(_tokenize_string("`%s`" % k))[1]
```
If this is still needed should convert to str format syntax

hwalinga replied:
OK

WillAyd reviewed (changes requested):
Sorry I know this is probably taking longer than you hope but there is a lot of ambiguity to `query` as is, so just really want to keep this as simple and succinct as possible. Some new comments to that effect

hwalinga commented:
@WillAyd No need to apologize. I know pandas is a very popular Python package. So, it is good to take the time for this.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -489,16 +489,51 @@ def _get_index_resolvers(self):
+            for toknum, tokval in _tokenize_string(expr)
+            if toknum == _BACKTICK_QUOTED_STRING
+        }
+        cleaned_names_expr = (
```
Can you use a set difference instead of another loop?

WillAyd replied:
I would rather lose that information and keep the implementation and corresponding error message much simpler

hwalinga replied:
I need that information to know if I need to raise the error message. I need to know if there is name with count > 1. If that is the case, I need to raise the error message. If not, I don't have to. 

However, I have changed the iterator comprehension to a map / filter combination. Maybe that is a bit more readable than the `name in name for ... if name in ...`.

hwalinga replied:
Take a look at the new commits.

WillAyd commented on pandas/core/generic.py:
```diff
@@ -489,16 +489,51 @@ def _get_index_resolvers(self):
+            name for name in cleaned_names if name in backtick_variables_expr
+        )
+        # Raise the ValueError when a name is counted more than once in the keys.
+        for name, count in collections.Counter(cleaned_names_expr).most_common(1):
```
So I appreciate how detailed of an error message this is but would prefer to just raise a simple error message (maybe even without the detail of all o the columns) and document this limitation elsewhere. The docstring for `query` would work

hwalinga replied:
OK

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -39,10 +36,125 @@
+        # "#": "_HASH_",
+        **{
+            char: "_" + token.tok_name[tokval] + "_"
+            # The ignore here is because of a bug in mypy that is resolved in 0.720
```
We are using mypy 0.72 in CI - can this now be deleted?

hwalinga replied:
OK

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -39,10 +36,125 @@
+            char: "_" + token.tok_name[tokval] + "_"
+            # The ignore here is because of a bug in mypy that is resolved in 0.720
+            for char, tokval in tokenize.EXACT_TOKEN_TYPES.items()  # type: ignore
+            if len(char) == 1
```
Hmm I thought from a previous review we didn't need to limit this to just one character?

hwalinga replied:
Well, the point is, multiple character operators are automatically replaced by single character operators. So multiple character operators *will* be replaced. However, for performance reasons, we don't have to check the string for multiple character operators. So, I just remove them here, as they will be replaced anyway.

WillAyd replied:
Cool thanks for the clarification. Unless there is a tangible performance benefit would rather keep the implementation simple

hwalinga replied:
Well, there is also a factor of undefined behavior. Officially the ordering of the dict is not guaranteed in Python version < 3.7, which means that if there is a multiple character operator inside the string there are randomized two outcomes: Are the multiple character operator is replaced by the multiple character itself, or it is replaced by the individual characters it is made off. 

Very, very unlikely to go wrong. IIRC the CPython implementation already does it, but there are no guarantees, so I rather have this line in place instead.

WillAyd replied:
Doesn't the iteration on line 89 go character by character anyway? So wouldn't match on multi character keys

hwalinga replied:
You are right. I forgot how exactly I implemented the character replacement and made wrong assumptions about it. In that case, we can just remove this line.

WillAyd commented on pandas/core/computation/expr.py:
```diff
@@ -39,10 +36,125 @@
+        '"': "_DOUBLEQUOTE_",
+        # Currently not possible. Terminates parser and won't find backtick.
+        # "#": "_HASH_",
+        **{
```
Can you create this dict separately and then just call `special_characters_replacements.update(...` on it? I think easier to read than the argument unpacking here

WillAyd commented on pandas/core/generic.py:
```diff
@@ -489,16 +489,43 @@ def _get_index_resolvers(self):
+        )
+        for name, count in collections.Counter(cleaned_names_in_expr).most_common(1):
+            if count > 1:
+                raise ValueError("Description {} is ambiguous.".format(name))
```
I think needs a test that hits this

hwalinga replied:
That is covered by pandas/tests/frame/test_query_eval.py::TestDataFrameQueryBacktickQuoting::test_ambiguity_exception

If you take a look in the automated tests, you can also see I achieve 100 % coverage.

WillAyd commented on doc/source/whatsnew/v0.25.2.rst:
```diff
@@ -101,6 +101,7 @@ Other
 ^^^^^
 
 - Compatibility with Python 3.8 in :meth:`DataFrame.query` (:issue:`27261`)
+- Backtick quoting in :meth:`DataFrame.query` and :meth:`DataFrame.eval` can now also be used to use invalid identifiers like names that start with a digit, are python keywords, or are using single character operators.
```
Can you move this to the 1.0.0 whatsnew?

hwalinga replied:
OK

WillAyd reviewed (changes requested):
Thanks this is looking a lot simpler. cc @jreback for thoughts as well

hwalinga commented:
Well, I had some doubt over that filter / map thingy. Usually the comprehensions are favored over the those, and the `__contains__` isn't so nice either. But for people that are a bit more familiar with functional programming I think it reads a lot better than the previous version. 
    
I have changed the doc, but I will wait for the push. (Don't want to put too much work on your automated pipelines.) As I expect to still have some more improvements and probably also squash the commits etc.

jreback reviewed (changes requested):
i'll have to review, but I am not sure I am on-board with the rationale behind this. .query doesn't need to grow a completely different DSL than python, this makes maintenance very burdensome. you can *always* use quoted strings w/o query to do this, so I am leaning towards not expanding this scope.

hwalinga commented:
@jreback 

This PR does not only add the functionality of using special characters in backtick quoted identifiers. It improves on a wide range of issues with my latest PR (#24955). 

To summarize, this PR consists of the following improvements:

* Raise an error when it is not possible to convert a backtick quoted string to a valid Python identifier. (These cases are rare, but it is good to still raise an understandable error.) (`_create_valid_python_identifier`)
* Raise an error when the conversion of the column resolvers to valid Python identifiers creates an ambiguity. This has to do with how Python parses spaces. This issue was already in there with the previous PR, but wasn't handled. Although I doubt someone in the world would ever differentiate its column names based on a difference in whitespace, it is still good to have such an error raised, as it would otherwise pass silently. (`_get_space_character_free_column_resolvers`)
* Moved all code from common.py to expr.py. (I think it is better to have that at one place, I realized.)
* The actual improvement of using special characters in the backtick quoted strings. (`_remove_spaces_column_name` -> `_create_valid_python_identifier`)
* Tests to cover the above.
* Improved documentation that also covers the edge cases.

I think the PR improves on a lot of things, and not only adding more coverage for different cases for the backtick quoted strings. 

Also, keep in mind that we are not expanding the DSL for the .query function. We are only adding more coverage for different cases in naming, i.e. special characters, next to spaces.

hwalinga commented:
Yet another improvement. I was able to fix the problem with the multiple spaces. I just made use of the indices of the token_generator to extract the exact indices of the backtick quoted string. So, you completely circumvent any undesired manipulation done to the tokens. 

This will simplify the PR also because it removes documenting that edge case and catching an error when that occurs (that was a significant portion of code.).

hwalinga commented:
@jreback Can you do the codereview?

Current failed test if just some linting in an unrelated file.

jreback commented on pandas/core/computation/expr.py:
```diff
@@ -38,10 +34,107 @@
 
 import pandas.io.formats.printing as printing
 
+# A token value Python's tokenizer probably will never use.
```
rather than do this here, pls make another module, maybe pandas/core/compuatation/tokenize.py

hwalinga replied:
`tokenize` is already a Python module. Should we call it `query_parsing`?

jreback replied:
maybe call it parsing

jreback commented on pandas/core/computation/eval.py:
```diff
@@ -160,7 +160,7 @@ def _check_for_locals(expr: str, stack_level: int, parser: str):
 
     if at_top_of_stack or not_pandas_parser:
-        for toknum, tokval in tokenize_string(expr):
+        for toknum, tokval in _tokenize_string(expr):
```
see my comments below, we don't want to use private functions

hwalinga replied:
What exactly do you mean with the comments below? Do you mean moving the functions to a new module?

jreback replied:
I mean don't use private functions (so don't privatize it in the first place)

jreback commented on pandas/core/frame.py:
```diff
@@ -3101,8 +3101,10 @@ def query(self, expr, inplace=False, **kwargs):
 
-            You can refer to column names that contain spaces by surrounding
-            them in backticks.
+            You can refer to column names that contain spaces or operators by
```
likely this would need to have a versioncahnged 1.0.0 here as well

hwalinga replied:
Should I specify what changed in the versions like this: 

```
            You can refer to column names that contain spaces or operators by
            surrounding them in backticks. This way you can also escape
            names that start with a digit, or are a Python keyword. Basically
            when it is not valid Python identifier.

            For example, if one of your columns is called ``a a`` and you want
            to sum it with ``b``, your query should be ```a a` + b``.

            .. versionadded:: 0.25.0: Backtick quoting introduced.

            .. versionadded:: 1.0.0: Expanding functionality for more than only spaces.
```

jreback replied:
sure

jreback commented on pandas/tests/frame/test_query_eval.py:
```diff
@@ -1091,7 +1104,56 @@ def test_mixed_underscores_and_spaces(self, df):
         res = df.eval("A + `C_C`")
         expect = df["A"] + df["C_C"]
         tm.assert_series_equal(res, expect)
+
```
can you create a new test class (and move all of the quoting testst here)

hwalinga replied:
All my tests regarding quoting are already in one class: TestDataFrameQueryBacktickQuoting

jreback replied:
k great

jreback commented on pandas/core/computation/parsing.py:
```diff
@@ -0,0 +1,154 @@
+    return name
+
+
+def tokenize_and_clean_column_names(name: Union[str, int]) -> Union[str, int]:
```
why would this take/return an int?

hwalinga replied:
This function is called for all column names regardless of they are being used in the query. For column names, pandas will accept integers as well. That's why this function must also accept integers, so that for the following example, query works as well. 

```python
import pandas as pd
df = pd.DataFrame({0: [1, 2, 3], "A A": [2, 3, 4]})
df.query("`A A` > 3")
```

hwalinga replied:
As a name resolver with integers is worthless, I removed the inter names before being inserted into the dictionary. This will simplify some type hints, and fallback code for integer names.

jreback commented on pandas/core/computation/parsing.py:
```diff
@@ -0,0 +1,154 @@
+    token_generator = tokenize.generate_tokens(line_reader)
+
+    # Loop over all tokens till a backtick (`) is found.
+    # Then, take all tokens till the next backtick to form a backtick quoted
```
what happens if you only have 1 backtick? SyntaxError? is this tested?

hwalinga replied:
You will get an unreferenced variable error. I can catch this and re-throw a more readable ValueError. (SyntaxError is only for actual Python code.)

hwalinga replied:
Changed to SyntaxError as that is used elsewhere in the pandas code as well.

jreback commented on pandas/core/frame.py:
```diff
@@ -3161,6 +3167,21 @@ def query(self, expr, inplace=False, **kwargs):
+        This can lead to the following problem.
+
+        The parsing can result in an error for certain characters.
+        These characters are those that fall outside the ASCII range
```
are there sufficient examples below?

hwalinga replied:
I looked a bit more in depth to this, and made it a bit more detailed. I also looked into the lexical analysis document of Python and I think there are no more difficult cases as described here.

jreback commented on pandas/core/frame.py:
```diff
@@ -3161,6 +3167,21 @@ def query(self, expr, inplace=False, **kwargs):
+        (U+0001..U+007F), but also the hashtag (as it used for comments)
+        and unpaired quotes (unpaired brackets are fine).
+
+        If you encounter such a problem you can fall back to square bracket notation.
```
remove this last paragraph

hwalinga replied:
OK

jreback commented on pandas/core/frame.py:
```diff
@@ -3161,6 +3167,21 @@ def query(self, expr, inplace=False, **kwargs):
+
+        The parsing can result in an error for certain characters.
+        These characters are those that fall outside the ASCII range
+        (U+0001..U+007F), but also the hashtag (as it used for comments)
```
are there tests for when backquoating *doesn't* work?

hwalinga replied:
There was one, but I added two more.

jreback commented on pandas/core/frame.py:
```diff
@@ -3161,6 +3167,21 @@ def query(self, expr, inplace=False, **kwargs):
 
+        *Backtick quoted variables*
+
+        Backtick quoted variables are parsed as literal Python code and
```
can you add a reference here to the python identifier docs

hwalinga replied:
OK

jreback commented on pandas/tests/frame/test_query_eval.py:
```diff
@@ -1051,8 +1051,21 @@ def df(self):
                 "A": [1, 2, 3],
                 "B B": [3, 2, 1],
                 "C C": [4, 5, 6],
+                "C  C": [7, 4, 3],
```
can you add a doc-string to this fixture, saying these are all valid backtick quoted strings

hwalinga replied:
OK

jreback reviewed (changes requested):
conceptually ok with this change

jreback commented:
@pandas-dev/pandas-core if others can have a look here

datapythonista commented:
Sorry if this is a silly question @hwalinga, I understand the goal of this is to be able to run something like:
```python
import pandas
df = pandas.DataFrame({'$ amount': [10, 200, 20]})
df.query('`$ amount` < 50')
```
And what you're doing is converting `$ amount < 50` to `_DOLLARSIGN__amount > 50` before the `eval` is executed.

Why don't we execute the `eval` as `locals()['$ amount'] < 50` instead?

hwalinga commented:
@datapythonista 

An interesting observation; I never thought about that. However, upon trying it, I got the an error saying that the function `locals` is not supported by the `query` function. 

I don't know the exact workings of the downstream code, but I think such a solution in which we try to support locals for downstream code might become too complex, also because there are different engines by which the query can be executed. There is python and numexpr, and I don't know what will happen in the future. If more engines will be added we need to support locals for that as well, while this solution will work regardless. 

Therefor, in my view, albeit the current solution is a bit hacky, it does provide a simpler query string for any downstream code working with it.

jreback commented:
> @datapythonista
> 
> An interesting observation; I never thought about that. However, upon trying it, I got the an error saying that the function `locals` is not supported by the `query` function.
> 
> I don't know the exact workings of the downstream code, but I think such a solution in which we try to support locals for downstream code might become too complex, also because there are different engines by which the query can be executed. There is python and numexpr, and I don't know what will happen in the future. If more engines will be added we need to support locals for that as well, while this solution will work regardless.
> 
> Therefor, in my view, albeit the current solution is a bit hacky, it does provide a simpler query string for any downstream code working with it.

@hwalinga so the point is this. we *don't* need to actually use numexpr here for non-trivial usages. we can simply execute it directly (this in fact is what ``engine='python'``. I am all for reorganizing the code, but I agree it would be nice to actually simplify this.

hwalinga commented:
So, the proposal is to move the name resolving to the ast tree instead (the parsing used by engine='python'). The ast parser is a very complicated piece of code and there is already a lot of custom stuff in there. The tokenization is much easier and there is not so much custom stuff in that. I think it is not a good idea to move some of the functionality to the ast parser, and here I summed up why:

 * I don't think it is a good idea to bring more customization to the already complex tree parser.

 * I think it is better to keep the code regarding backtick quoting clustered and not to scatter it around with some in the token parsing and some in the tree parsing. I like that is now all in one place in the `parsing.py` file.

 * Backtick quoting is more related to the token parsing. The code now just simplifies the query before it got send to the engine. (It resolves the backtick quoting, and updates the name resolvers.)

 * I think pandas should simplify the query the most as possible before it sends it to the engine, because the engine could be a third party engine. (Like the 'numexpr'.) This will probably also help in the future whenever there is the possibility to add more engines. You want to keep the amount of features of the query that they are processing to a minimum, so that those engines can focus primarily on the processing than on the parsing. So, they only deal with a mostly simple query in addition to the name resolvers, not too much other stuff.

 * Numexpr does not support locals()[]. We can use such a thing only for our own custom parser and that will make this feature only available for engine='python'. This might be not a very big of a deal (most people probably just use the python engine and don't bother with installing numexpr), and I agree we don't have to write overly complicated code to support both. However, we already support both with the current implementation and since 'numexpr' is the default engine it also would not really make sense to not support that one. (It however autofallbacks to python, which I also don't think is a very good idea, but I elaborated on that already in #30005)

 * This would also mean a regression, as 25.3 already supports backtick quoting for both numexpr and python already.

With all this in mind, I think we should just proceed with the current implementation we have now.

jbrockmendel commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -656,6 +656,7 @@ Other
 - :meth:`Series.append` will no longer raise a ``TypeError`` when passed a tuple of ``Series`` (:issue:`28410`)
 - :meth:`SeriesGroupBy.value_counts` will be able to handle the case even when the :class:`Grouper` makes empty groups (:issue: 28479)
 - Fix corrupted error message when calling ``pandas.libs._json.encode()`` on a 0d array (:issue:`18878`)
+- Backtick quoting in :meth:`DataFrame.query` and :meth:`DataFrame.eval` can now also be used to use invalid identifiers like names that start with a digit, are python keywords, or are using single character operators. (:issue:`27017`)
```
is pd.eval affected?

hwalinga replied:
No, DataFrame.eval makes use of pd.eval, but that is downstream in de code of the changes made in this PR. pd.eval expects a query and some name resolvers. This PR interprets the backtick quotes and replaces them with Python valid identifiers in addition to updating the name resolvers, before sending them downstream and then eventually to pd.eval. But pd.eval is not affected by this.

jbrockmendel commented on pandas/core/computation/parsing.py:
```diff
@@ -0,0 +1,159 @@
+
+
+def create_valid_python_identifier(name: str) -> str:
+    """Create valid Python identifiers from any string.
```
newline after triple quotes

hwalinga replied:
Well, this is not what PEP 257 suggested. I don't think it is very strict, but I also see both styles in the pandas code base, so I picked the one suggested in the PEP. 

https://www.python.org/dev/peps/pep-0257/

hwalinga replied:
Well, this is not what PEP 257 suggested. I don't think it is very strict, but I also see both styles in the pandas code base, so I picked the one suggested in the PEP. 

https://www.python.org/dev/peps/pep-0257/

hwalinga replied:
Okay, changed it to that now.

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -486,21 +486,24 @@ def _get_axis_resolvers(self, axis):
-        """Return the space character free column resolvers of a dataframe.
+        return {tokenize_and_clean_column_names(k) or k: v for k, v in d.items()}
+
+    def _get_special_character_free_column_resolvers(self) -> dict:
```
can the return type be more specific?

hwalinga replied:
Yes ... done.

jbrockmendel commented:
> the possibility to add more engines

Is this something we actually anticipate happening?

hwalinga commented:
@jbrockmendel 

I think we don't expect new engines to be added, but with this statement in the readme `Additionally, it has the broader goal of becoming the most powerful and flexible open source data analysis / manipulation tool available in any language.`, I thought query might play a role in that. So, that is just one of the reasons I wanted to keep the query that is send to the engine as simple as possible.

hwalinga commented:
Test failures unrelated to this PR.

simonjayhawkins commented:
> Test failures unrelated to this PR.

if you merge master these should be fixed

hwalinga commented:
Thanks, I thought that would be done automatically. My bad.

hwalinga commented:
Another unrelated failed test:

```
pyarrow.lib.ArrowInvalid: Could not convert NA with type NAType: tried to convert to boolean
```

jorisvandenbossche commented:
> Another unrelated failed test:

That should be fixed in master

jreback commented:
@hwalinga responses inline

> * I don't think it is a good idea to bring more customization to the already complex tree parser.

actually the ast parser is what python uses, i am not sure we are expanding this *at* all.

> * I think it is better to keep the code regarding backtick quoting clustered and not to scatter it around with some in the token parsing and some in the tree parsing. I like that is now all in one place in the `parsing.py` file.

yes I agree with this, your code-org is nice

> * Backtick quoting is more related to the token parsing. The code now just simplifies the query before it got send to the engine. (It resolves the backtick quoting, and updates the name resolvers.)
> * I think pandas should simplify the query the most as possible before it sends it to the engine, because the engine could be a third party engine. (Like the 'numexpr'.) This will probably also help in the future whenever there is the possibility to add more engines. You want to keep the amount of features of the query that they are processing to a minimum, so that those engines can focus primarily on the processing than on the parsing. So, they only deal with a mostly simple query in addition to the name resolvers, not too much other stuff.

this is true, but you are missing the points. the *only* engine that is supported is ``numexpr`` which must have strings. I don't think in *pandas* we will *ever* have another engine, The simple soln is simply to fallback to actual python execution where columns names can be *anything*.

> * Numexpr does not support locals()[]. We can use such a thing only for our own custom parser and that will make this feature only available for engine='python'. This might be not a very big of a deal (most people probably just use the python engine and don't bother with installing numexpr), and I agree we don't have to write overly complicated code to support both. However, we already support both with the current implementation and since 'numexpr' is the default engine it also would not really make sense to not support that one. (It however autofallbacks to python, which I also don't think is a very good idea, but I elaborated on that already in #30005)

this has nothing to do with numexpr, all variable references are resolved prior to sending to the engine

> * This would also mean a regression, as 25.3 already supports backtick quoting for both numexpr and python already.

no it wouldn't, this is an extension of parsing.

jreback commented:
@hwalinga 

so I think your code cleanup is +1. I don't really think supporting special characters aside from space (e.g. ```$```) are worth it, but its not much code so ok.

I would like to see a follown, that handles a SyntaxError in the numexpr engine and then just falls back to the python engine.

jreback commented:
@jbrockmendel @jorisvandenbossche @datapythonista comments.

WillAyd reviewed (approved):
Updating approval per previous comments but lgtm

jreback commented:
@hwalinga can you merge master

hwalinga commented:
@jreback I have merged master and squashed the commits. There are some build problems on the 32 bit Linux, and some problems with building the docs. Seems unrelated.

hwalinga commented:
@jreback All is green. I think you can merge this now.

hwalinga commented:
Hold it. Something seems to be gone wrong in squashing the commit.

hwalinga commented:
@jreback Everything correct now, ready for merge.

jreback commented:
can you merge master and ping on green.

hwalinga commented:
@jreback Yes, all green, squashed, ready for the merge.

jreback commented on pandas/core/computation/eval.py:
```diff
@@ -12,7 +12,8 @@
 
 from pandas.core.computation.engines import _engines
-from pandas.core.computation.expr import Expr, _parsers, tokenize_string
+from pandas.core.computation.expr import Expr, _parsers
```
we should de-privatize parsers (and change it to a more informative, parser_engines, but should be a followup)

hwalinga replied:
#30681

jreback commented on pandas/core/computation/parsing.py:
```diff
@@ -0,0 +1,157 @@
+    return toknum, tokval
+
+
+def clean_column_names(name: str) -> str:
```
can you call this: clean_column_name (not the plural) to replace its action

jreback commented on pandas/core/computation/parsing.py:
```diff
@@ -0,0 +1,157 @@
+def tokenize_backtick_quoted_string(
+    token_generator: Iterator[tokenize.TokenInfo], source: str, string_start: int
+) -> Tok:
+    """
```
can you add Parameters / Returns here

jreback commented on pandas/core/computation/parsing.py:
```diff
@@ -0,0 +1,157 @@
+
+def tokenize_backtick_quoted_string(
+    token_generator: Iterator[tokenize.TokenInfo], source: str, string_start: int
+) -> Tok:
```
this return annotation is not correct, its a Tuple[....] (be specific inside the tuple)

hwalinga replied:
I created `Tok` as an alias for `Tuple[int, str]`, but as this did not seem clear, I made them all explicit.

jreback commented on pandas/core/computation/parsing.py:
```diff
@@ -0,0 +1,157 @@
+    ----------
+    source : str
+        A Python source code string
+    """
```
can you add Returns

jreback commented on pandas/core/generic.py:
```diff
@@ -453,22 +453,29 @@ def _get_axis_resolvers(self, axis):
 
-    def _get_index_resolvers(self):
-        d = {}
+    def _get_index_resolvers(self) -> Dict[str, ABCSeries]:
```
as a followup I'd like to clean up all of these functions; i think we can move them entirely to pandas/computation/ (including the axis resolvers which is just plain wrong now the way its implemented), can you create an issue for this

hwalinga replied:
#30683

simonjayhawkins replied:
and replace ABCSeries in type annotations. ABCSeries resolves to any.

hwalinga replied:
Replace with what? Also, let's continue this in #30683

simonjayhawkins replied:
I _think_ return type is Dict[str, Union["Series", MultiIndex]]

simonjayhawkins replied:
> Also, let's continue this in #30683

I don't think typing discussion relevant there

hwalinga replied:
Okay, I will see if that works. 

The issue is meant for cleaning up these functions. So, why not fix the typing hints as well.

jreback reviewed (changes requested):
some comments, pls rebase, ping on green and we can get this in.

hwalinga committed:
ENH: Add ability to use special characters for column names in query function.

Clean up in the code that is used for using spaces in the query function
and extending the ability to also use special characters that are
not allowed in python identifiers.

All files related to this functionality are now in the
pandas/core/computation/parsing.py file.

hwalinga commented:
@jreback 

Improved docstrings and rebased.

jreback reviewed (approved):
thanks @hwalinga i know this has taken a while!

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -454,22 +454,29 @@ def _get_axis_resolvers(self, axis):
 
-    def _get_space_character_free_column_resolvers(self):
-        """Return the space character free column resolvers of a dataframe.
+        return {clean_column_name(k): v for k, v in d.items() if k is not int}
```
should this be `if not isinstance(k, int)`?  this is causing mypy failures in #30694.

jreback replied:
yes that looks like it needs changing

hwalinga replied:
Oeh, that's really embarrassing. Yes, it should be `if not isinstance(k, int)`.

What should be best to fix this. Should I make a PR for this?

hwalinga replied:
#30698
